File: 3-HARD/Python/8-regular-expression-matching-dp/proposed-solution/regex_matcher.py
class RegexMatcher:
    def __init__(self, text: str, pattern: str):
        self.text = text
        self.pattern = pattern
        self.dp = [[False] * (len(pattern) + 1) for _ in range(len(text) + 1)]

    def is_match(self) -> bool:
        self.dp[0][0] = True

        for j in range(2, len(self.pattern) + 1):
            if self.pattern[j - 1] == '*':
                self.dp[0][j] = self.dp[0][j - 2]

        def matches(i, j):
            if self.pattern[j - 1] == '.' or self.pattern[j - 1] == self.text[i - 1]:
                return self.dp[i - 1][j - 1]
            if self.pattern[j - 1] == '*':
                zero = self.dp[i][j - 2]
                one_or_more = (
                    (self.pattern[j - 2] == '.' or self.pattern[j - 2] == self.text[i - 1])
                    and self.dp[i - 1][j]
                )
                return zero or one_or_more
            return False

        for i in range(1, len(self.text) + 1):
            for j in range(1, len(self.pattern) + 1):
                self.dp[i][j] = matches(i, j)

        return self.dp[len(self.text)][len(self.pattern)]

File: 3-HARD/JavaScript/8-regular-expression-matching-dp/proposed-solution/RegexMatcher.js
// RegexMatcher.js - Regular Expression Matching using Dynamic Programming in JavaScript

class RegexMatcher {
  constructor(text, pattern) {
    this.text = text;
    this.pattern = pattern;
    this.dp = Array.from({ length: text.length + 1 }, () =>
      new Array(pattern.length + 1).fill(false)
    );
  }

  isMatch() {
    this.dp[0][0] = true;

    for (let j = 2; j <= this.pattern.length; j++) {
      if (this.pattern[j - 1] === "*") {
        this.dp[0][j] = this.dp[0][j - 2];
      }
    }

    const matches = (i, j) => {
      if (
        this.pattern[j - 1] === "." ||
        this.pattern[j - 1] === this.text[i - 1]
      ) {
        return this.dp[i - 1][j - 1];
      }
      if (this.pattern[j - 1] === "*") {
        const zero = this.dp[i][j - 2];
        const oneOrMore =
          (this.pattern[j - 2] === "." ||
            this.pattern[j - 2] === this.text[i - 1]) &&
          this.dp[i - 1][j];
        return zero || oneOrMore;
      }
      return false;
    };

    for (let i = 1; i <= this.text.length; i++) {
      for (let j = 1; j <= this.pattern.length; j++) {
        this.dp[i][j] = matches(i, j);
      }
    }

    return this.dp[this.text.length][this.pattern.length];
  }
}

export default RegexMatcher;
